File: weisi/src/main/java/com/cjq/lib/weisi/util/SimpleReflection.java
package com.cjq.lib.weisi.util;

import android.support.annotation.NonNull;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Created by CJQ on 2018/3/28.
 */

public class SimpleReflection {

    private SimpleReflection() {
    }

    public static Type getInterfaceParameterizedType(@NonNull Object interfaceInstance, @NonNull Class interfaceClass, int interfaceTypeParameterPosition) {
        Class c = interfaceInstance.getClass();
        do {
            Type[] interfaceTypes = c.getGenericInterfaces();
            if (interfaceTypes == null) {
                throw new IllegalArgumentException("interfaceInstance may implement an interface");
            }
            Type interfaceType;
            for (int i = 0;i < interfaceTypes.length;++i) {
                interfaceType = interfaceTypes[i];
                if (interfaceType instanceof ParameterizedType) {
                    ParameterizedType parameterizedType = (ParameterizedType) interfaceType;
                    if (parameterizedType.getRawType() == interfaceClass) {
                        if (interfaceTypeParameterPosition < 0 || interfaceTypeParameterPosition >= parameterizedType.getActualTypeArguments().length) {
                            throw new IllegalArgumentException("interface type parameter position out of bounds");
                        }
                        return parameterizedType.getActualTypeArguments()[interfaceTypeParameterPosition];
                    }
                }
            }
            c = c.getSuperclass();
        } while (!c.isInstance(Object.class));
        throw new IllegalArgumentException("no proper parameter");
    }

    public static Type getClassParameterizedType(@NonNull Object classInstance, @NonNull Class classClass, int classTypeParameterPosition) {
        Class c = classInstance.getClass();
        do {
            Type interfaceType = c.getGenericSuperclass();
            if (interfaceType instanceof ParameterizedType) {
                ParameterizedType parameterizedType = (ParameterizedType) interfaceType;
                if (classTypeParameterPosition < 0 || classTypeParameterPosition >= parameterizedType.getActualTypeArguments().length) {
                    throw new IllegalArgumentException("class type parameter position out of bounds");
                }
                return parameterizedType.getActualTypeArguments()[classTypeParameterPosition];
            }
            c = c.getSuperclass();
        } while (!c.isInstance(Object.class));
        throw new IllegalArgumentException("no proper parameter");
    }
}

File: wsb/src/main/java/com/wsn/lib/wsb/util/SimpleReflection.kt
package com.wsn.lib.wsb.util

import com.sun.istack.internal.NotNull

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * Created by CJQ on 2018/3/28.
 */

object SimpleReflection {

    @JvmStatic
    fun getInterfaceParameterizedType(interfaceInstance: Any, interfaceClass: Class<*>, interfaceTypeParameterPosition: Int): Type {
        var c: Class<*> = interfaceInstance.javaClass
        do {
            val interfaceTypes = c.genericInterfaces
                    ?: throw IllegalArgumentException("interfaceInstance may implement an interface")
            var interfaceType: Type
            for (i in interfaceTypes.indices) {
                interfaceType = interfaceTypes[i]
                if (interfaceType is ParameterizedType) {
                    val parameterizedType = interfaceType
                    if (parameterizedType.rawType === interfaceClass) {
                        if (interfaceTypeParameterPosition < 0 || interfaceTypeParameterPosition >= parameterizedType.actualTypeArguments.size) {
                            throw IllegalArgumentException("interface type parameter position out of bounds")
                        }
                        return parameterizedType.actualTypeArguments[interfaceTypeParameterPosition]
                    }
                }
            }
            c = c.superclass
        } while (!c.isInstance(Any::class.java))
        throw IllegalArgumentException("no proper parameter")
    }

    @JvmStatic
    fun getClassParameterizedType(classInstance: Any, classTypeParameterPosition: Int): Type {
        var c: Class<*> = classInstance.javaClass
        do {
            val interfaceType = c.genericSuperclass
            if (interfaceType is ParameterizedType) {
                if (classTypeParameterPosition < 0 || classTypeParameterPosition >= interfaceType.actualTypeArguments.size) {
                    throw IllegalArgumentException("class type parameter position out of bounds")
                }
                return interfaceType.actualTypeArguments[classTypeParameterPosition]
            }
            c = c.superclass
        } while (!c.isInstance(Any::class.java))
        throw IllegalArgumentException("no proper parameter")
    }
}
